File: frontend/src/lib/core/api/layout.ts
import { createClient } from '@connectrpc/connect';
import { createConnectTransport } from '@connectrpc/connect-web';
import {
	ComponentType,
	LayoutService,
	type Component as ProtoComponent,
	type Layout as ProtoLayout,
	type LayoutTile as ProtoLayoutTile,
	type Panel as ProtoPanel
} from '$lib/gen/layout/v1/layout_pb.js';

import { timestampToIso } from './connect';
import { API_BASE, api } from './client';

export interface Layout {
	id: string;
	project_id: string;
	name: string;
	total_panels: number;
	total_capacity_kw: number;
	tile_count: number;
	created_at: string;
	updated_at: string;
}

export interface LayoutTile {
	id: string;
	layout_id: string;
	bbox: BoundingBox;
	lod_level: number;
	panel_count: number;
	metadata_json: string;
}

export interface Panel {
	id: string;
	tile_id: string;
	string_id: string;
	geometry_geojson: string;
	tilt: number;
	azimuth: number;
	elevation: number;
}

export interface BoundingBox {
	min_x: number;
	min_y: number;
	max_x: number;
	max_y: number;
}

export interface PanelArrayParams {
	panel_width: number;
	panel_height: number;
	tilt_angle: number;
	azimuth: number;
	row_spacing: number;
	column_spacing: number;
	panel_asset_id?: string;
	panel_model?: string;
	panel_rated_power_w?: number;
	fill_area_geojson: string;
	terrain_layer_id?: string;
}

function isUuidLike(value: string | undefined): boolean {
	if (!value) return false;
	return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(value);
}

function finiteOr(value: number, fallback: number): number {
	return Number.isFinite(value) ? value : fallback;
}

export interface Component {
	id: string;
	layout_id: string;
	asset_id: string;
	component_type: string;
	position: { longitude: number; latitude: number; elevation: number };
	rotation: number;
	metadata_json: string;
}

export interface ZoneAllocation {
	zone_type: string;
	target_area_sqm: number;
	planned_area_sqm: number;
	coverage_percent: number;
	geometry_geojson: string;
	design_description: string;
}

export interface ZonePlanResult {
	layout_id: string;
	boundary_area_sqm: number;
	target_capacity_mw: number;
	estimated_dc_mw: number;
	recommended_ac_mw: number;
	zones: ZoneAllocation[];
	assumptions: string[];
}

const LAYOUT_API_BASE =
	import.meta.env.VITE_LAYOUT_API_BASE_URL ||
	(API_BASE.endsWith(':8080') ? API_BASE.replace(':8080', ':8082') : API_BASE);

const transport = createConnectTransport({
	baseUrl: LAYOUT_API_BASE
});

const client = createClient(LayoutService, transport);

function componentTypeFromProto(componentType: ComponentType): string {
	switch (componentType) {
		case ComponentType.PANEL:
			return 'panel';
		case ComponentType.INVERTER:
			return 'inverter';
		case ComponentType.TRANSFORMER:
			return 'transformer';
		case ComponentType.TRACKER:
			return 'tracker';
		case ComponentType.SUBSTATION:
			return 'substation';
		case ComponentType.COMBINER_BOX:
			return 'combiner_box';
		case ComponentType.JUNCTION_BOX:
			return 'combiner_box';
		default:
			return 'panel';
	}
}

function componentTypeToProto(componentType?: string): ComponentType {
	switch (componentType) {
		case 'panel':
			return ComponentType.PANEL;
		case 'inverter':
			return ComponentType.INVERTER;
		case 'transformer':
			return ComponentType.TRANSFORMER;
		case 'tracker':
			return ComponentType.TRACKER;
		case 'substation':
			return ComponentType.SUBSTATION;
		case 'combiner_box':
		case 'junction_box':
			return ComponentType.COMBINER_BOX;
		default:
			return ComponentType.UNSPECIFIED;
	}
}

function mapLayout(layout?: ProtoLayout): Layout {
	if (!layout) {
		throw new Error('Layout response was empty');
	}

	return {
		id: layout.id,
		project_id: layout.projectId,
		name: layout.name,
		total_panels: layout.totalPanels,
		total_capacity_kw: layout.totalCapacityKw,
		tile_count: layout.tileCount,
		created_at: timestampToIso(layout.createdAt),
		updated_at: timestampToIso(layout.updatedAt)
	};
}

function mapTile(tile?: ProtoLayoutTile): LayoutTile {
	if (!tile) {
		throw new Error('Layout tile response was empty');
	}

	return {
		id: tile.id,
		layout_id: tile.layoutId,
		bbox: {
			min_x: tile.bbox?.minX ?? 0,
			min_y: tile.bbox?.minY ?? 0,
			max_x: tile.bbox?.maxX ?? 0,
			max_y: tile.bbox?.maxY ?? 0
		},
		lod_level: tile.lodLevel,
		panel_count: tile.panelCount,
		metadata_json: tile.metadataJson
	};
}

function mapPanel(panel?: ProtoPanel): Panel {
	if (!panel) {
		throw new Error('Panel response was empty');
	}

	return {
		id: panel.id,
		tile_id: panel.tileId,
		string_id: panel.stringId,
		geometry_geojson: panel.geometryGeojson,
		tilt: panel.tilt,
		azimuth: panel.azimuth,
		elevation: panel.elevation
	};
}

function mapComponent(component?: ProtoComponent): Component {
	if (!component) {
		throw new Error('Component response was empty');
	}

	return {
		id: component.id,
		layout_id: component.layoutId,
		asset_id: component.assetId,
		component_type: componentTypeFromProto(component.componentType),
		position: {
			longitude: component.position?.longitude ?? 0,
			latitude: component.position?.latitude ?? 0,
			elevation: component.position?.elevation ?? 0
		},
		rotation: component.rotation,
		metadata_json: component.metadataJson
	};
}

function parsePolygonCoordinates(fillAreaGeojson: string): [number, number][] | null {
	try {
		const parsed = JSON.parse(fillAreaGeojson) as {
			type?: string;
			coordinates?: number[][][];
		};
		if (parsed.type !== 'Polygon' || !parsed.coordinates?.[0]?.length) return null;
		return parsed.coordinates[0].map((c) => [Number(c[0]), Number(c[1])]);
	} catch {
		return null;
	}
}

function isLikelyLonLat(coords: [number, number][]): boolean {
	if (coords.length === 0) return false;
	return coords.every(([lon, lat]) => Math.abs(lon) <= 180 && Math.abs(lat) <= 90);
}

function centroidLatitude(coords: [number, number][]): number {
	const sum = coords.reduce((acc, [, lat]) => acc + lat, 0);
	return sum / coords.length;
}

function meterToDegreeScale(meters: number, latDeg: number): number {
	const degPerMeterLat = 1 / 111320;
	const cosLat = Math.max(0.1, Math.cos((latDeg * Math.PI) / 180));
	const degPerMeterLon = 1 / (111320 * cosLat);

	// Use isotropic average to keep dimensions consistent in lon/lat plane.
	const degPerMeter = (degPerMeterLat + degPerMeterLon) / 2;
	return meters * degPerMeter;
}

function normalizeGenerationParams(params: PanelArrayParams): PanelArrayParams {
	const coords = parsePolygonCoordinates(params.fill_area_geojson);
	if (!coords || !isLikelyLonLat(coords)) return params;

	const lat = centroidLatitude(coords);

	// The backend uses row_spacing and column_spacing as the full centre-to-centre
	// pitch for the grid sweep, not as the inter-panel gap.
	// Convert: pitch = panel_dimension + gap, then to degrees.
	const colPitchDeg = meterToDegreeScale(params.panel_width + params.column_spacing, lat);
	// When row_spacing is 0 the backend auto-computes it from tilt; keep it 0 so
	// that logic fires.  Otherwise send the full row pitch (panel height + gap).
	const rowPitchDeg = params.row_spacing > 0
		? meterToDegreeScale(params.panel_height + params.row_spacing, lat)
		: 0;

	return {
		...params,
		panel_width: meterToDegreeScale(params.panel_width, lat),
		panel_height: meterToDegreeScale(params.panel_height, lat),
		row_spacing: rowPitchDeg,
		column_spacing: colPitchDeg
	};
}

export const layoutApi = {
	create: async (projectId: string, name: string) => {
		const response = await client.createLayout({ projectId, name });
		return { layout: mapLayout(response.layout) };
	},

	get: async (id: string) => {
		const response = await client.getLayout({ id });
		return { layout: mapLayout(response.layout) };
	},

	list: async (projectId: string) => {
		const response = await client.listLayouts({ projectId });
		return { layouts: response.layouts.map(mapLayout) };
	},

	delete: async (id: string) => {
		await client.deleteLayout({ id });
		return {};
	},

	generatePanelArray: async (layoutId: string, params: PanelArrayParams) => {
		const normalized = normalizeGenerationParams(params);
		let fillAreaPayload: unknown = normalized.fill_area_geojson;
		try {
			fillAreaPayload = JSON.parse(normalized.fill_area_geojson);
		} catch {
			fillAreaPayload = normalized.fill_area_geojson;
		}
		const body: Record<string, unknown> = {
			panel_width: finiteOr(normalized.panel_width, 0),
			panel_height: finiteOr(normalized.panel_height, 0),
			tilt_angle: finiteOr(normalized.tilt_angle, 20),
			azimuth: finiteOr(normalized.azimuth, 180),
			row_spacing: finiteOr(normalized.row_spacing, 0),
			column_spacing: finiteOr(normalized.column_spacing, 0),
			fill_area_geojson: fillAreaPayload
		};

		if (normalized.panel_asset_id && normalized.panel_asset_id.trim().length > 0) {
			body.panel_asset_id = normalized.panel_asset_id;
		}
		if (normalized.panel_model && normalized.panel_model.trim().length > 0) {
			body.panel_model = normalized.panel_model;
		}
		if ((normalized.panel_rated_power_w ?? 0) > 0) {
			body.panel_rated_power_w = normalized.panel_rated_power_w;
		}
		if (isUuidLike(normalized.terrain_layer_id)) {
			body.terrain_layer_id = normalized.terrain_layer_id;
		}

		const response = await api.post<{
			total_panels: number;
			tile_count: number;
			total_capacity_kw: number;
		}>(`/api/v1/layouts/${layoutId}/generate-array`, body);

		return {
			panels_created: response.total_panels,
			tiles_created: response.tile_count,
			capacity_kw: response.total_capacity_kw
		};
	},

	planZones: async (layoutId: string, boundaryGeojson: string, targetCapacityMw: number) => {
		return api.post<ZonePlanResult>(`/api/v1/layouts/${layoutId}/plan-zones`, {
			boundary_geojson: boundaryGeojson,
			target_capacity_mw: targetCapacityMw
		});
	},

	getTiles: async (layoutId: string, viewport: BoundingBox, lodLevel: number) => {
		const response = await client.getTiles({
			layoutId,
			viewport: {
				minX: viewport.min_x,
				minY: viewport.min_y,
				maxX: viewport.max_x,
				maxY: viewport.max_y
			},
			lodLevel
		});

		return { tiles: response.tiles.map(mapTile) };
	},

	getTilePanels: async (tileId: string) => {
		const response = await client.getTilePanels({ tileId });
		return { panels: response.panels.map(mapPanel) };
	},

	placeComponent: async (layoutId: string, component: Partial<Component>) => {
		const response = await client.placeComponent({
			layoutId,
			assetId: component.asset_id ?? '',
			componentType: componentTypeToProto(component.component_type),
			position: {
				longitude: component.position?.longitude ?? 0,
				latitude: component.position?.latitude ?? 0,
				elevation: component.position?.elevation ?? 0
			},
			rotation: component.rotation ?? 0,
			metadataJson: component.metadata_json ?? ''
		});

		return { component: mapComponent(response.component) };
	},

	moveComponent: async (componentId: string, position: { longitude: number; latitude: number; elevation: number }, rotation = 0) => {
		const response = await client.moveComponent({
			componentId,
			position,
			rotation
		});
		return { component: mapComponent(response.component) };
	},

	removeComponent: async (componentId: string) => {
		await client.removeComponent({ componentId });
		return {};
	},

	listComponents: async (layoutId: string) => {
		const response = await client.listComponents({ layoutId });
		return { components: response.components.map(mapComponent) };
	}
};

File: mobile/lib/core/models/layout.dart
import 'package:equatable/equatable.dart';

class Layout extends Equatable {
  final String id;
  final String projectId;
  final String name;
  final int totalPanels;
  final double totalCapacityKw;
  final int tileCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Layout({
    required this.id,
    required this.projectId,
    required this.name,
    required this.totalPanels,
    required this.totalCapacityKw,
    required this.tileCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Layout.fromJson(Map<String, dynamic> json) {
    return Layout(
      id: json['id'] as String,
      projectId: json['project_id'] as String,
      name: json['name'] as String,
      totalPanels: json['total_panels'] as int? ?? 0,
      totalCapacityKw: (json['total_capacity_kw'] as num?)?.toDouble() ?? 0,
      tileCount: json['tile_count'] as int? ?? 0,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'project_id': projectId,
        'name': name,
        'total_panels': totalPanels,
        'total_capacity_kw': totalCapacityKw,
        'tile_count': tileCount,
      };

  @override
  List<Object?> get props => [id, projectId, name, totalPanels];
}

class BoundingBox extends Equatable {
  final double minX;
  final double minY;
  final double maxX;
  final double maxY;

  const BoundingBox({
    required this.minX,
    required this.minY,
    required this.maxX,
    required this.maxY,
  });

  factory BoundingBox.fromJson(Map<String, dynamic> json) {
    return BoundingBox(
      minX: (json['min_x'] as num).toDouble(),
      minY: (json['min_y'] as num).toDouble(),
      maxX: (json['max_x'] as num).toDouble(),
      maxY: (json['max_y'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
        'min_x': minX,
        'min_y': minY,
        'max_x': maxX,
        'max_y': maxY,
      };

  @override
  List<Object?> get props => [minX, minY, maxX, maxY];
}

class LayoutTile extends Equatable {
  final String id;
  final String layoutId;
  final BoundingBox bbox;
  final int lodLevel;
  final int panelCount;

  const LayoutTile({
    required this.id,
    required this.layoutId,
    required this.bbox,
    required this.lodLevel,
    required this.panelCount,
  });

  factory LayoutTile.fromJson(Map<String, dynamic> json) {
    return LayoutTile(
      id: json['id'] as String,
      layoutId: json['layout_id'] as String,
      bbox: BoundingBox.fromJson(json['bbox'] as Map<String, dynamic>),
      lodLevel: json['lod_level'] as int? ?? 0,
      panelCount: json['panel_count'] as int? ?? 0,
    );
  }

  @override
  List<Object?> get props => [id, layoutId, lodLevel];
}

class Panel extends Equatable {
  final String id;
  final String tileId;
  final String? stringId;
  final String geometryGeojson;
  final double tilt;
  final double azimuth;
  final double elevation;

  const Panel({
    required this.id,
    required this.tileId,
    this.stringId,
    required this.geometryGeojson,
    required this.tilt,
    required this.azimuth,
    required this.elevation,
  });

  factory Panel.fromJson(Map<String, dynamic> json) {
    return Panel(
      id: json['id'] as String,
      tileId: json['tile_id'] as String,
      stringId: json['string_id'] as String?,
      geometryGeojson: json['geometry_geojson'] as String,
      tilt: (json['tilt'] as num).toDouble(),
      azimuth: (json['azimuth'] as num).toDouble(),
      elevation: (json['elevation'] as num).toDouble(),
    );
  }

  @override
  List<Object?> get props => [id, tileId];
}

class PanelArrayParams {
  final double panelWidth;
  final double panelHeight;
  final double tiltAngle;
  final double azimuth;
  final double rowSpacing;
  final double columnSpacing;
  final String fillAreaGeojson;
  final String? terrainLayerId;

  const PanelArrayParams({
    required this.panelWidth,
    required this.panelHeight,
    required this.tiltAngle,
    required this.azimuth,
    required this.rowSpacing,
    required this.columnSpacing,
    required this.fillAreaGeojson,
    this.terrainLayerId,
  });

  Map<String, dynamic> toJson() => {
        'panel_width': panelWidth,
        'panel_height': panelHeight,
        'tilt_angle': tiltAngle,
        'azimuth': azimuth,
        'row_spacing': rowSpacing,
        'column_spacing': columnSpacing,
        'fill_area_geojson': fillAreaGeojson,
        if (terrainLayerId != null) 'terrain_layer_id': terrainLayerId,
      };
}

class Component extends Equatable {
  final String id;
  final String layoutId;
  final String? assetId;
  final String componentType;
  final ComponentPosition position;
  final double rotation;
  final String? metadataJson;

  const Component({
    required this.id,
    required this.layoutId,
    this.assetId,
    required this.componentType,
    required this.position,
    this.rotation = 0,
    this.metadataJson,
  });

  factory Component.fromJson(Map<String, dynamic> json) {
    return Component(
      id: json['id'] as String,
      layoutId: json['layout_id'] as String,
      assetId: json['asset_id'] as String?,
      componentType: json['component_type'] as String,
      position: ComponentPosition.fromJson(
          json['position'] as Map<String, dynamic>),
      rotation: (json['rotation'] as num?)?.toDouble() ?? 0,
      metadataJson: json['metadata_json'] as String?,
    );
  }

  @override
  List<Object?> get props => [id, layoutId, componentType];
}

class ComponentPosition extends Equatable {
  final double longitude;
  final double latitude;
  final double elevation;

  const ComponentPosition({
    required this.longitude,
    required this.latitude,
    this.elevation = 0,
  });

  factory ComponentPosition.fromJson(Map<String, dynamic> json) {
    return ComponentPosition(
      longitude: (json['longitude'] as num).toDouble(),
      latitude: (json['latitude'] as num).toDouble(),
      elevation: (json['elevation'] as num?)?.toDouble() ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
        'longitude': longitude,
        'latitude': latitude,
        'elevation': elevation,
      };

  @override
  List<Object?> get props => [longitude, latitude, elevation];
}
